addons/mail_client_extension/controllers/mail_client_extension.py

```diff
@@ -14,32 +15,159 @@
+                                                     ('email_normalized', '=', normalized_email)], limit=1)
+
+        response = self._prepare_contact_values(partner)
+
+        # if no partner is found in the database, we should also return an empty one having id = -1, otherwise older versions of
+        # plugin won't work
+        if not response['partner']:
+            sender_domain = normalized_email.split('@')[1]
+            response['partner'] = {
+                'id': -1,
+                'email': email,
+                'name': name,
+                'enrichment_info': None,
+            }
+            company = self._find_existing_company(normalized_email)
+            if not company:  # create and enrich company
+                company, enrichment_info = self._create_company_from_iap(sender_domain)
+                response['partner']['enrichment_info'] = enrichment_info
+            response['partner']['company'] = self._prepare_company_values(company)
+
+        return response
+
+    @http.route('/mail_client_extension/partners/search', type="json", auth="outlook", cors="*")
+    def res_partners_search(self, query, limit=30, **kwargs):
+        """
+        Used for the plugin search contact functionality where the user types a string query in order to search for
+        matching contacts, the string query can either be the name of the contact, it's reference or it's email.
+        We choose these fields because these are probably the most interesting fields that the user can perform a
+        search on.
+        The method returns an array containing the dicts of the matched contacts.
+        """
+
+        #In a multi-company environment, the method may return contacts not belonging to the company that the user
+        #is connected to, this may result in the user not being able to view the contact in Odoo, while this may happen
+        #it is not supported for now and users are encouraged to check if they are connected to the correct company before
+        # clicking on a contact.
+        filter_domain = ['|', '|', ('display_name', 'ilike', query), ('ref', '=', query), ('email', 'ilike', query)]
+        # Search for the partner based on the email.
+        # If multiple are found, take the first one.
+        partners = request.env['res.partner'].search(filter_domain, limit=limit)
+
+        partners = [
+            self._prepare_partner_values(partner)
+            for partner in partners
+        ]
+        return {"partners": partners}
+
+    @http.route('/mail_client_extension/partner/create', type="json", auth="outlook", cors="*")
+    def res_partner_create(self, email, name, company, **kwargs):
+        # TODO search the company again instead of relying on the one provided here?
+        # Create the partner if needed.
+        partner_info = {
+            'name': name,
+            'email': email,
+        }
+        if company and company > -1:
```

Comment:
Mmmh ... you should call it parent_company_id, if we really support an id coming from the outside world. Could be great to check it exists also (and cast as an int).

Reply:
Yeah indeed, but since this is a request param, this would break older versions of the plugin :/

Reply:
Yeah, I've added a docstring for that.